consensus/istanbul/backend/backend.go

```diff
@@ -255,6 +256,30 @@ func (sb *Backend) ParentBlockValidators(proposal istanbul.Proposal) istanbul.Va
 	return sb.getOrderedValidators(proposal.Number().Uint64()-1, proposal.ParentHash())
 }
 
+func (sb *Backend) NextBlockValidators(proposal istanbul.Proposal) (istanbul.ValidatorSet, error) {
+	state, err := sb.stateAt(proposal.ParentHash())
+	if err != nil {
+		return nil, err
+	}
+	state = state.Copy()
+
+	istExtra, err := types.ExtractIstanbulExtra(proposal.Header())
+	if err != nil {
+		return nil, err
+	}
+
+	// There was no change
+	if len(istExtra.AddedValidators) == 0 && istExtra.RemovedValidators.BitLen() == 0 {
+		return sb.ParentBlockValidators(proposal), nil
+	}
+
+	newValSet, err := sb.getNewValidatorSet(proposal.Header(), state)
```

Comment:
It seems that this code already implicitly trusts the header input, so it may be better to use the snapshot functions to construct the validator set here. Namely by calling `snap := sb.snapshot(...)` to get the snapshot for the parent of the proposal and using `snap.apply(...)` to add the validator set from the given proposal header. `getNewValidatorSet` calls into a very expensive EVM call (it uses more than the block gas limit).

https://github.com/celo-org/celo-blockchain/blob/4ea7566836aee9cdbaa6c4561c3f41b9d79037c7/consensus/istanbul/backend/engine.go#L673
https://github.com/celo-org/celo-blockchain/blob/4ea7566836aee9cdbaa6c4561c3f41b9d79037c7/consensus/istanbul/backend/snapshot.go#L94
https://github.com/celo-org/celo-blockchain/blob/4ea7566836aee9cdbaa6c4561c3f41b9d79037c7/params/protocol_params.go#L164

Reply:
`apply` wasn't usable for me since it requires an `Epoch` long jump.
I did utilize the method I learnt from it though!